src/os/README.md

```diff
@@ -50,19 +50,20 @@ inspiration for several decisions in this library.
 - do not use `OS\ErrnoException` if the error condition would not be indicated
   by the `errno` variable in C. Consider adding another similar class, e.g.
   add `OS\HErrnoException` if you want to report an error exposed via `h_errno`
-- add and use Hack classes (not type aliases) for 'handle'-like parameters and
-  return values, e.g. `OS\open()` returns a `HH\Lib\FileDescriptor` instead of
-  an `int`; as well as aiding type safety, this prevents requests from interfering
-  with resources that belong to another request.
+- add and use Hack classes (not type aliases) for long-lived 'handle'-like
+  parameters and return values, e.g. `OS\open()` returns a
+  `HH\Lib\FileDescriptor` instead of an `int`; as well as aiding type safety,
+  this prevents requests from interfering with resources that belong to another
+  request.
 - Avoid `inout` parameters; return tuples instead. For example, prefer
   `function mkstemp(string $pattern): (FileDescriptor, string)` to
   `function mkstemp(inout string $in_pattern_out_path): FileDescriptor`
   - this can aid for common use by allowing string literals, rather than
     requiring otherwise-unused locals
-  - if the primary purpose of a function is to mutate its parameters, this
-    should be considered case-by-case; for example, if we were to add something
-    similar to `array_pop()`, it probably should take an `inout` container - but
-    probably shouldn't be in `OS\`
+  - if the primary purpose of a set of functions is to create, mutate and
+    destroy a short-lived C pointer, they should be exposed as a `shape`
+    or `vec`. See [Appendix: short-lived C pointer
+    encoding](#appendix-short-lived-c-pointer-encoding) section for more detail.
```

Comment:
this is still only applicable for some short-lived pointers: ones that represent lists, or hidden structs.

If we wanted a general rule for C pointers, it would have to be 'one class per pointer type/usage, no vecs/shape' - this would mostly avoid semantic understanding, but it would need to understand which subsystem is using the pointer, and how/if to free them.

The vec case could perhaps be covered with: if a short-lived C pointer and related functions are used to emulate a datastructure that does not natively exist in C but does in Hack, use the corresponding Hack datastructure instead. For example, if a pointer acts like an ordered list, use a `vec`.

I'm not sure how to generalize the shape case (and we need to consider if `OS\inaddr` and friends  are a mistake, or if there's a good reason to treat them differently)

Reply:
> If we wanted a general rule for C pointers, it would have to be 'one class per pointer type/usage, no vecs/shape' - this would mostly avoid semantic understanding, but it would need to understand which subsystem is using the pointer, and how/if to free them.

Apparently we have no choice but create a closeable Hack class wrappers for C native resources, and this PR does not discuss those native resources. What are in question are those C value types  or borrowing types, i.e.the OS would not hold a reference to the C data type. We have three option to deal with them:

1. Create closeable Hack class wrappers for them, consistent with C native resources but not ease to use.
2. Define Hack value types, i.e. `shape` or `vec`, in the `OS` namespace, and only create short-lived the C native data internally on demand.
3. Define garbage-collectable Hack classes in the `OS` namespace, and only create short-lived the C native data internally on demand.

I don't have a strong opinion among these options, but currently `OS\in_addr` is designed per option 2.

Reply:
I think this fits with https://github.com/hhvm/hsl/discussions/169#discussioncomment-1816963 :)